main/src/ca/uwaterloo/flix/api/lsp/provider/completion/MagicMatchCompleter.scala

```diff
@@ -36,78 +36,164 @@ object MagicMatchCompleter {
     }
   }
 
   /**
-    * Creates a case string and its corresponding right-hand side string.
+    * Formats the cases of an enum into a string.
     */
-  private def createCase(sym: Symbol.CaseSym, cas: TypedAst.Case, z: Int): (String, String, Int) = {
-    cas.tpe.typeConstructor match {
-      case Some(TypeConstructor.Unit) =>
-        (s"$sym", s"$${${z + 1}:???}", z + 1)
-      case Some(TypeConstructor.Tuple(arity)) =>
-        val elements = List.range(1, arity + 1).map(i => s"$${${i + z}:_elem$i}").mkString(", ")
-        (s"$sym($elements)", s"$${${arity + z + 1}:???}", z + arity + 1)
-      case _ =>
-        (s"$sym($${${z + 1}:_elem})", s"$${${z + 2}:???}", z + 2)
+  private def mkEnumMatchBody(cases: Map[Symbol.CaseSym, TypedAst.Case]): String = {
+    val maxCaseLength = cases.values.map(getCaseLength).max
+    val sb = new StringBuilder
+    var index = 1
+    for ((sym, cas) <- cases.toList.sortBy(_._1.loc)) {
+      val oldIndex = index
+      val (lhs, newZ) = createCase(sym, cas, index)
+      val paddedLhs = padLhs(lhs, maxCaseLength, oldIndex, newZ)
+      sb.append(s"    case $paddedLhs => $${$index:???}\n")
+      index = newZ + 1
     }
+    sb.toString()
   }
 
   /**
-    * Formats the cases of an enum into a string.
+    * Convert the given type into a list of Member.
+    * If the given type is an enum, return a list of EnumCase.
+    * Otherwise, return a list of SingleCase, which contains only a string indicating the idx of the member.
     */
-  private def patternMatchBody(cases: Map[Symbol.CaseSym, TypedAst.Case]): String = {
-    val maxCaseLength = cases.values.map(getCaseLength).max
+  private def type2member(tpe: Type, idx: Int)(implicit root: TypedAst.Root): List[Member] =
+    getEnumSym(tpe) match {
+      case Some(sym) => root.enums(sym).cases.toList.map{case (sym, cas) => EnumCase(sym, cas)}
+      case None => SingleCase(s"_member$idx") :: Nil
+    }
+
+  /**
+    * Returns the cartesian product of the given list of lists.
+    * The length of the returned list is the product of the lengths of the input lists.
+    *
+    * Example:
+    *  Given [ [Red, Green], [Circle, Square] ]
+    *  The cartesian product is [ [Red, Circle], [Red, Square], [Green, Circle], [Green, Square] ]
+    */
+  private def cartesianProduct(casesPerType: List[List[Member]]): List[List[Member]] = {
+    casesPerType.foldLeft(List(List.empty[Member])) { (acc, list) =>
+      for {
+        x <- acc
+        y <- list
+      } yield x :+ y
+    }
+  }
+
+  /**
+    * Formats the cases of a tuple into a string.
+    */
+  private def mkTupleMatchBody(members: List[List[Member]]): String = {
     val sb = new StringBuilder
-    var z = 1
-    for ((sym, cas) <- cases.toList.sortBy(_._1.loc)) {
-      val (lhs, rhs, newZ) = createCase(sym, cas, z)
-      val paddedLhs = padLhs(lhs, maxCaseLength)
-      sb.append(s"    case $paddedLhs => $rhs\n")
-      z = newZ
+    val maxCaseLength = members.map(getMemberLength).max
```

Comment:
Do we know this .max cannot cause an exception?

Reply:
The .max will not cause an exception as long as the length of the tuple is not zero.

Is it possible for us to create an empty tuple in flix?

Reply:
No. So that should be good.